src/main/java/week5/AlphabetFinder.java

```diff
@@ -0,0 +1,83 @@
+package week5;
+
+
+import static java.util.stream.Collectors.toSet;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+
+public class AlphabetFinder {
+
+  /*
+   * The graph is represented using adjacency List as 1) Alphabet might be unbounded, 2)
+   * from example the graph seem to be sparse 3) we do not check the presence of arbitrary edges
+   * but systematically traverse through the graph.
+   *
+   * Main Steps:
+   *  1. Build partial order graph of characters from word list.
+   *  2. Topological sort to generate a total order.
+   *
+   * Returns : empty <- if there is a cycle in the words
+   *           List<Character> <- the ordered alphabet
+   *
+   * Time Complexity: O(build) + O(sort)
+   *                : O(n*m*len) + O(|alphabet| + len)
+   *                : O(n*m*len)
+   *
+   * where
+   *    n is the number of words,
+   *    len is the len of words,
+   *    m is the fan-out of words
+   */
+  public static Optional<List<Character>> findAlphabet(final String[] words) {
+
+    if (words == null) {
+      return Optional.empty();
+    }
+
+    return buildGraph(words).topologicalSort();
+  }
+
+
+  /*
+   * The while loop iteratively groups the word according to the first character.
+   * The list of first characters are inserted into the graph and each group of suffixes
+   * is added to the queue to infer its partial orders.
+   *
+   * Time Complexity: O(n*m*len)
+   * where
+   *    n is the number of words,
+   *    len is the len of words,
+   *    m is the fan-out of words
+   */
+  private static Graph buildGraph(String[] words) {
+
+    Set<Character> nodes =
+        Arrays.stream(words)
+            .flatMapToInt(String::chars)
+            .mapToObj(s -> (char) s)
+            .collect(toSet());
+
+    Graph graph = new Graph(nodes);
+
+    for (int i = 1; i < words.length; i++) {
+      String word1 = words[i - 1];
+      String word2 = words[i];
+
+      getEdgeFrom(word1, word2).ifPresent(chars -> graph.addEdge(chars));
+    }
+
+    return graph;
+  }
+
+  private static Optional<char[]> getEdgeFrom(String word1, String word2) {
+    for (int i = 0; i < Math.min(word1.length(), word2.length()); i++) {
```

Comment:
The optimiser *may* spot this, but better to pull the constant expression out of the loop and have e.g. "i < len"